packages/mui-base/src/unstable_useNumberInput/useNumberInput.ts

```diff
@@ -105,6 +158,14 @@ export function useNumberInput(parameters: UseNumberInputParameters): UseNumberI
     }
   }, [formControlContext, disabledProp, focused, onBlur]);
 
+  React.useEffect(() => {
+    if (isControlled && isNumber(value)) {
+      dispatch({
+        type: NumberInputActionTypes.resetInputValue,
+      });
+    }
+  }, [value, dispatch, isControlled]);
```

Comment:
@sai6855 I've adapted your fix in https://github.com/mui/material-ui/pull/40348 here, as this PR will replace the manual `setState`s

I think it shouldn't be necessary to check for potential `string`s because it's not allowed for the `value`, and TS/PropTypes can take care of that. Also when it's uncontrolled, the `inputValue` should not ever get out of sync with the `value` like in https://github.com/mui/material-ui/issues/40340

For this, I've added an additional `resetInputValue` action CC @michaldudak